DOC-655 Docs for updated asset selection syntax

neverett opened the pull request:
## Summary & Motivation

Updates "Asset selection syntax" doc with the new asset selection syntax.

## How I Tested These Changes

Local build.

## Changelog

> Insert changelog entry or delete this section.

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax-1-9-11.md:
```diff
@@ -0,0 +1,355 @@
+
+## Syntax usage
+
+A query includes a list of clauses. Clauses are separated by commas, except in the case of the `selection` parameter of the following methods. In these cases, each clause is a separate element in a list:
```
no commas in the new syntax

neverett commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -1,351 +1,112 @@
 ---
-title: 'Asset selection syntax'
+title: "Asset selection syntax"
```
@salazarm sorry, the previous iteration of this PR was a bit confusing -- this is the new asset selection syntax doc.

salazarm commented on docs/src/code-examples-content.js:
```diff
@@ -1,610 +0,0 @@
-
```
deleting this seems unintentional ?

neverett replied:
This file is rebuilt in production, so doesn't need to be checked in (although checking it in won't cause any harm)

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
````diff
@@ -85,19 +67,19 @@ def manhattan_stats(database: DuckDBResource):
 
 ```python
-manhattan_job = define_asset_job(name="manhattan_job", selection="manhattan/manhattan_stats")
+manhattan_stats = AssetSelection.from_string('key:"manhattan/manhattan_stats"')
````
this can be 
```suggestion
manhattan_stats = define_asset_job(name="manhattan_job", selection='key:"manhattan/manhattan_stats"')
```

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
````diff
@@ -1,58 +1,42 @@
 
 ```python
-raw_data_job = define_asset_job(name="raw_data_job", selection="taxi_zones_file")
+taxi_zones_file = AssetSelection.from_string('key:"taxi_zones_file"')
````
same

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -1,58 +1,42 @@
-## Examples
+### Select a single asset \{#single-asset}
 
-The examples in this section use the following asset graph from the [Dagster University Essentials project](https://github.com/dagster-io/project-dagster-university) to demonstrate how to use the selection syntax:
-
-![Screenshot of Daggy U project graph](/images/guides/build/assets/asset-selection-syntax/asset-selection-syntax-dag.png)
-
```
Should we use the Daggy U project graph as an example?

neverett replied:
I was wondering that myself -- I think Daggy U is on dagster v1.9, so the project graph would be OK, but the asset selection interface wouldn't have the new syntax. We might need different examples for the ones that were carried over from the old asset selection syntax doc that used Daggy U / taxi trips data.

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -109,26 +91,26 @@ manhattan/manhattan_stats
-raw_data_job = define_asset_job(
-    name="taxi_zones_job", selection=["taxi_zones_file", "taxi_trips_file"]
-)
+taxi_zones_and_trips = AssetSelection.from_string("taxi_zones_file, taxi_trips_file")
```
```suggestion
raw_data_job = define_asset_job(
    name="taxi_zones_job", selection="key:taxi_zones_file or key:taxi_trips_file"
)
```

salazarm replied:
>TODO how would you do this in new syntax?

Check the suggestion, you need to use ` or `

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -150,102 +132,116 @@ taxi_zones_file taxi_trips_file
 <TabItem value="cli" label="CLI">
 
-When selecting an asset's entire lineage using the CLI, enclose the asterisk (`*`) and the asset key in double quotes (`"`):
+When selecting an asset's dependencies using the CLI, enclose the plus sign (`+`) and the asset key in double quotes (`"`):
```
`lineage` is correct because it includes both dependencies and dependents of the asset.

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -150,102 +132,116 @@ taxi_zones_file taxi_trips_file
+### Select a specific number of upstream layers \{#specific-upstream}
 
-To select an asset and all its upstream dependencies, add an asterisk (`*`) before the asset key in the query.
+To select an asset and multiple upstream layers, add a number and plus sign (`+`) for the layers you want to select before the asset key in the query.
 
-This example selects the `manhattan_map` asset and all its upstream dependencies.
+This example selects the `manhattan_map` asset and two upstream layers.
```
Hmm, this also seems redundant given the table above 🤔 . Maybe we move the table the bottom as a reference?

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
````diff
@@ -150,102 +132,116 @@ taxi_zones_file taxi_trips_file
 
 ```python
-manhattan_job = define_asset_job(name="manhattan_job", selection="*manhattan_map")
+manhattan_two_upstream = AssetSelection.from_string("2+manhattan_map")
````
same, this can continue using `define_asset_job`.

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
````diff
@@ -254,98 +250,281 @@ manhattan_job = define_asset_job(name="manhattan_job", selection="++manhattan_ma
 
 ```shell
-taxi_zones_file*
+taxi_zones_file++
````
```suggestion
taxi_zones_file+
```

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -254,98 +250,281 @@ manhattan_job = define_asset_job(name="manhattan_job", selection="++manhattan_ma
+| Filter | Syntax | Description |
+|--------|--------|-------------|
+| **Exact key** | `key:"my_key"` | Selects assets with the exact key `my_key`. |
+| **Key substring** | `key_substring:"partial_key"` | Selects assets whose key contains `partial_key`. |
```
We remove `key_substring`, instead we're allowing wildcards for the `key` filter only.

For example:

`key:prefix_*` would select all assets with keys starting with `prefix_`. 
You can also have multiple wildcards:

`key:prefix_*_middlefix_*_suffix*`

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -254,98 +250,281 @@ manhattan_job = define_asset_job(name="manhattan_job", selection="++manhattan_ma
+| **Group** | `group:"team1"` | Selects assets in the group `team1`. |
+| **Kind** |  `kind:"table"` |  Selects assets of kind `table`. |
+| **Code location** | `code_location:"repo1"` | Selects assets located in code location `repo1`. |
+
```
In Dagster Plus we additionally support the following filters:

`column_tag`
`columns`
`table_name`

salazarm replied:
In the Gantt chart we also support a selection syntax:
<img width="542" alt="Screenshot 2025-02-20 at 1 52 57 PM" src="https://github.com/user-attachments/assets/41e29a17-bc96-4be9-9437-fd8a891e5eb5" />

This one only supports the attributes: `name`, `status`

salazarm replied:
https://elementl.dagster.cloud/prod/runs/3ca30bf9-e2f3-4dad-8e8e-7c44159a792c

<img width="1728" alt="Screenshot 2025-02-20 at 1 54 53 PM" src="https://github.com/user-attachments/assets/70367a85-96d2-47e1-95a2-13f371fb0b69" />

salazarm replied:
Full list in cloud:
<img width="814" alt="Screenshot 2025-02-20 at 10 02 45 AM" src="https://github.com/user-attachments/assets/a81a8cd1-ae69-4848-8b6e-74cb9428ef29" />


Additionally, in branch deployments we support the `changed_in_branch` filter, however we don't support `column`, `column_tag`, `table` in branch deployments.

We could add another column to the table "Supported views". And for each attribute we can list "OSS", "Dagster Plus", "Branch deployments".

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
````diff
@@ -254,98 +250,281 @@ manhattan_job = define_asset_job(name="manhattan_job", selection="++manhattan_ma
+### Select upstream and downstream assets with filters \{#filters}
+
+```shell
++key:"data_pipeline" and kind:"table"+
+```
+
+Selects one layer upstream and one layer downstream of `data_pipeline`, limited to assets of kind `table`.
````
For 1 layer this should be `1+key:"data_pipeline" and kind:"table"+1`

salazarm reviewed (changes requested):
Overall this looks good! Left comments inline but will also summarize some of the new things we want to add:
- Explaining that the `key` filter supports wildcard matching (the other filters do not)
- Explaining that some filters are only available in Dagster+ and one (`changed_in_branch`) is only available in Dagster+ branch deployments. And that some are only available in the Gantt chart (Step selection)

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -1,77 +1,94 @@
+| Filter | Syntax | Description | Supported views |
+|--------|--------|-------------|-----------------|
+| **Exact key** | `key:"my_key"` | Selects assets with the exact key `my_key`. | OSS, Dagster+, Dagster+ branch deployments |
+| **Key substring with wildcard filter** | `key:partial_key_*`| Selects assets whose key contains `partial_key`. | OSS, Dagster+, Dagster+ branch deployments |
```
```suggestion
| **Key with wildcard filter** | `key:partial_key_*`| Selects assets whose key starts with `partial_key_`. | OSS, Dagster+, Dagster+ branch deployments |
```

neverett replied:
Incorporated into my most recent set of changes

salazarm commented:
I want to find a place to also show these examples which a lot of our customers say has been the most useful capability added for them:

- Find paths between two assets:
  - `key:"start"+ and +key:"end"`
- Find paths between two sets of assets:
  -  `(key:"start-1" or key:"start-2")+ and +(key:"end-1" or key:"end-2")`
- Find paths between two assets that go through the "middle" asset:
  -  `key:"start"+ and +key:"middle"+ and +key:"end"`
- Find paths between two sets of assets by tag:
  - `tag:"private"+ and +tag:"public"

neverett commented:
@salazarm this is ready for another review -- here's what I changed recently:

- Updated introduction and "syntax" section (and incorporated feedback from earlier reviews).
- Added a `dagster_university` code location to dogfood so its [assets](https://dogfood-test-1.dogfood.dagster.cloud/prod/locations/dagster_university/assets) are available for examples and screenshots.
- Updated as many examples as possible to use the assets in `dagster_university` (and marked where I still need to take screenshots).
- Marked a few remaining examples as still TODO -- the assets in `dagster_university` don't really work for these examples, so I may just fill in the Python and CLI tabs and remove the UI tabs for now.

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -1,351 +1,440 @@
-| `ASSET_KEY*`          | Selects an asset and all of its downstream dependencies. [See an example](#all-downstream).                                                                                                                                                                                                                           |
-| `+ASSET_KEY`          | Selects an asset and one layer upstream of the asset. Including multiple `+`s will select that number of upstream layers from the asset. Any number of `+`s is supported. [See an example](#specific-upstream).    |
-| `ASSET_KEY+`          | Selects an asset and one layer downstream of the asset. Including multiple `+`s will select that number of downstream layers from the asset. Any number of `+`s is supported. [See an example](#specific-downstream). |
+- The Asset Catalog
+- The Alert Policy creation page (when creating an asset alert)
+- The Insights page
```
The Asset Health overview page.

salazarm replied:
And the same ones as OSS too (When reading it, it sounds like a disjoint set of surfaces from the OSS one)

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -1,351 +1,440 @@
+| **Column tag** | `column:tag: "my_tag"` | Selects assets tagged with `my_tag`. | Dagster+ only |
+| **Columns** | `columns: "my_column"` | Selects assets with a column named `my_column`. | Dagster+ only |
+| **Table name** | `table_name: "my_table"` | Selects assets with a table named `my_table`. | Dagster+ only |
+| **Changed in branch** | `changed_in_branch: "my_branch"` | Selects assets changed in a branch named `my_branch`. | Dagster+ branch deployments only |
```
change_in_branch is actually an ENUM.
It can have the following values:

```
export enum ChangeReason {
  CODE_VERSION = 'CODE_VERSION',
  DEPENDENCIES = 'DEPENDENCIES',
  METADATA = 'METADATA',
  NEW = 'NEW',
  PARTITIONS_DEFINITION = 'PARTITIONS_DEFINITION',
  REMOVED = 'REMOVED',
  TAGS = 'TAGS',
}
```

We'll likely add more over time

salazarm replied:
This allows you to filter for assets that changed (for a particular reason) in your branch deployment

salazarm replied:
Oh also "ANY" is a possible value (meaning find assets that changed for any reason in the branch deployment).

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax.md:
```diff
@@ -1,351 +1,440 @@
-
-<Tabs>
-<TabItem value="python" label="Python">
+| Function | Syntax | Description |
```
Maybe `Function | Example | Description` 

roots | `roots(owner:"team:billing")` | Find all root assets depended on by the billing team
sinks | `sinks(owner:"team:billing")` | Find all sink assets depending on assets by the billing team

salazarm commented:
Hmm, I kind of want the path examples to stand out a bit more, It's lost in the sea of examples and I think when people see it a lightbulb will go off on their head that maybe we should move it to the top?

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax/index.md:
```diff
@@ -0,0 +1,32 @@
+In the **Dagster+ UI**, the asset selection syntax is available on:
+
+- The Asset Catalog page
+- The global asset lineage page
```
```suggestion
- The Global asset lineage page
```

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax/index.md:
```diff
@@ -0,0 +1,32 @@
+In the **Dagster OSS UI**, the asset selection syntax is available on:
+
+- The Asset Catalog page
+- The global asset lineage page
```
```suggestion
- The Global asset lineage page
```

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax/examples.md:
```diff
@@ -0,0 +1,349 @@
+sidebar_position: 100
+---
+
+This page contains common example asset selection queries and their implementation in Python, CLI, and the Dagster UI. For a full explanation of the filters, layers, operands, and functions that you can use to construct your own queries, see "[Asset selection syntax reference](reference)".
```
This is the first page in the dropdown and immediately directs you to the second page, so maybe the reference should go first?

neverett replied:
Yeah, I can put the reference page first -- I put the examples page first in order to make the path examples as prominent as possible, but those should be easier to find now that they're at the top of the page, regardless of the page's position in the sidebar.

salazarm replied:
Yeah I like this better!

salazarm commented:
One thing missing in the reference is `*` = "All assets". eg: `sinks(*)` (give me the sinks of all assets)

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax/reference.md:
```diff
@@ -0,0 +1,95 @@
+- **Sink assets** are assets without any downstream dependencies (leaf nodes), which means they don't provide input to any other assets.
+- **Root assets** are assets without any upstream dependencies (root nodes), which means no assets provide input to them.
+
+| Function | Syntax | Description |
+|--------|-------------|-----------|
+| **`sinks(expr)`** | `sinks(owner:"team:billing")` | Find all sink assets depending on assets by the billing team. |
+| **`roots(expr)`** | `roots(owner:"team:billing")` | Find all root assets depended on by the billing team. |
```
From dogfooding on docs + feature people were a bit confused by what the functions did.

This is the feedback: https://www.notion.so/dagster/Selection-Syntax-Dogfooding-1a718b92e4628060a3bee29204548734?pvs=4#1ac18b92e462809a9cfbe8cd77516141.

We need to make it clear we're find the Sinks/Roots OF the expression, and not WITHIN the expression.

neverett replied:
Gotcha, that makes sense -- is this accurate?

| Example | Description |
|--------|--------------|
| `sinks(owner:"team:billing")` | Find all assets without any downstream dependencies that depend on assets owned by the billing team. |
| `roots(owner:"team:billing")` | Find all assets without any upstream dependencies that feed into assets owned by the billing team. |

(I'll also add `sinks(*)` and roots(*)` to that table.)

neverett replied:
I'll also add a note specifying that we're finding the sinks/roots of the expression, not within the expression.

salazarm commented on docs/docs/dagster-plus/features/asset-catalog/catalog-views.md:
```diff
@@ -21,45 +21,42 @@ Catalog views require **Organization Admin**, **Admin**, or **Editor** permissio
+4. If modifying the view, note that any active filters will automatically be included in the set of changes. You can also change the view's name, description, icon, and sharing settings.
+5. When finished, click **Save changes**.
 
 ![Screenshot of editing catalog views](/images/dagster-plus/features/asset-catalog/edit-catalog-view.png)
```
This screenshot needs to be updated. It's showing the old UI

salazarm reviewed (approved):
Thank you so much Nikki! this looks great. Just a couple remaining things I left comments on but otherwise this looks good to me!

salazarm commented:
<img width="1089" alt="Screenshot 2025-03-10 at 9 10 12 PM" src="https://github.com/user-attachments/assets/3175c781-2619-4bc8-ba21-c166dab8cc85" />
@neverett would it be possible to make the dagster UI tab first since it has the image examples?

neverett commented:
@salazarm made a few more changes tonight:
* Made the UI tab the first one for all examples
* Added `roots()` and `sinks()` examples to examples.md
* Misc small copyedits

salazarm commented:
@neverett Also, if its not too much work, would it be possible to retake the screenshots? We shipped [this change](https://dagsterlabs.slack.com/archives/C04PD3G2F7H/p1741110688971199) which should make the graphs smaller

salazarm commented:
<img width="482" alt="Screenshot 2025-03-11 at 9 41 32 PM" src="https://github.com/user-attachments/assets/509ac218-61b1-402b-bfcb-c1dd4d374653" />
Also in user settings you can toggle this off to hide the "page load time"
<img width="245" alt="Screenshot 2025-03-11 at 9 41 56 PM" src="https://github.com/user-attachments/assets/3d80684f-7cb0-4643-8e4c-9c01d5d61531" />

neverett commented:
> @neverett Also, if its not too much work, would it be possible to retake the screenshots? We shipped [this change](https://dagsterlabs.slack.com/archives/C04PD3G2F7H/p1741110688971199) which should make the graphs smaller

Sure thing, I can do that first thing tomorrow morning.

salazarm commented on docs/docs/guides/build/assets/asset-selection-syntax/reference.md:
```diff
@@ -0,0 +1,100 @@
+| `sinks(*)` | Select all assets without any downstream dependencies. |
+| `roots(group:"public_data")` | Selects root assets within the `public_data` group.|
+| `sinks(group:"public_data")` | Selects sink assets within the `public_data` group. |
+| `roots(+group:"public_data")` | Selects all root assets that feed into the `public_data` group, but do not belong to that group. |
+| `sinks(group:"public_data"+)` | Selects all sink assets that depend on assets in the `public_data` group, but do not belong to that group. |
```
```suggestion
| `roots(+group:"public_data")` | Selects all root assets that feed into the `public_data` group. |
| `sinks(group:"public_data"+)` | Selects all sink assets that depend on assets in the `public_data` group. |
```

salazarm replied:
The sink/roots could belong to the group.

salazarm replied:
I guess maybe the intent is to signal that they could potentially be outside of the group? We could say something like 

"Selects all sink assets that depend on assets in the `public_data` group, they may or may not belong to the `public_data` group themselves."
